src/compiler/checker.ts

```diff
@@ -4173,7 +4173,23 @@ namespace ts {
                     const isLate = isLateBindableName(name);
                     const isWellKnown = isComputedPropertyName(name) && isWellKnownSymbolSyntactically(name.expression);
                     if (!isLate && !isWellKnown && isComputedNonLiteralName(name)) {
-                        return anyType;
+                        const exprType = checkExpression((name as ComputedPropertyName).expression);
+                        let indexerType: Type;
+                        if (isTypeAssignableToKind(exprType, TypeFlags.NumberLike)) {
+                            indexerType = getIndexTypeOfType(parentType, IndexKind.Number) || getIndexTypeOfType(parentType, IndexKind.String);
+                        }
+                        else if (isTypeAssignableToKind(exprType, TypeFlags.StringLike)) {

```

Comment:
```ts
const named = Symbol("");
let { [named]: prop } = strIndexed;
void prop;``
```

today that is `string` ... we probably wanna keep it this way until we have a symbol indexer.